tiledb/api/src/range.rs

```diff
@@ -0,0 +1,841 @@
+        datatype: Datatype,
+        cell_val_num: CellValNum,
+    ) -> TileDBResult<()> {
+        match self {
+            Self::Single(svr) => svr.check_datatype(datatype)?,
+            Self::Multi(_) => {
+                return Err(Error::InvalidArgument(anyhow!(
+                    "Dimensions can not have a fixed cell val num > 1"
+                )));
+            }
+            Self::Var(vvr) => vvr.check_datatype(datatype)?,
+        }
+
+        match cell_val_num {
+            CellValNum::Fixed(cvn) => {
+                if cvn.get() > 1 {
+                    return Err(Error::InvalidArgument(anyhow!(
+                        "Invalid cell val number: {}",
+                        cvn.get()
+                    )));
+                }
+                if datatype == Datatype::StringAscii {
+                    return Err(Error::InvalidArgument(anyhow!(
+                        "StringAscii dimensions must be var sized."
+                    )));
+                }
+                if !matches!(self, Self::Single(_)) {
+                    return Err(Error::InvalidArgument(anyhow!(
+                        "Non-string dimensions must have a cell val num of 1."
+                    )));
+                }
+            }
+            CellValNum::Var => {
+                if datatype != Datatype::StringAscii {
+                    return Err(Error::InvalidArgument(anyhow!(
+                        "Dimensions of type {} must have a cell val num of 1",
+                        datatype
+                    )));
+                }
+                if !matches!(self, Self::Var(VarValueRange::UInt8(_, _))) {
+                    return Err(Error::InvalidArgument(anyhow!(
+                        "String dimensions must use VarValueRange::UInt8"
+                    )));
+                }
+            }
+        }
+
+        Ok(())
+    }
+}
+
+impl Debug for Range {
+    fn fmt(&self, f: &mut Formatter) -> FmtResult {
+        write!(f, "{}", json!(self))
+    }
+}
+
+macro_rules! range_single_from {
+    ($($V:ident : $U:ty),+) => {
+        $(
+            impl From<&[$U; 2]> for Range {
+                fn from(value: &[$U; 2]) -> Range {
+                    Range::Single(SingleValueRange::from(value))
+                }
+            }
+        )+
+    }
+}
+
+range_single_from!(UInt8: u8, UInt16: u16, UInt32: u32, UInt64: u64);
+range_single_from!(Int8: i8, Int16: i16, Int32: i32, Int64: i64);
+range_single_from!(Float32: f32, Float64: f64);
+
+macro_rules! range_multi_try_from {
+    ($($V:ident : $U:ty),+) => {
+        $(
+            impl TryFrom<(u32, Box<[$U]>, Box<[$U]>)> for Range {
+                type Error = crate::error::Error;
+                fn try_from(value: (u32, Box<[$U]>, Box<[$U]>)) -> TileDBResult<Range> {
+                    Ok(Range::Multi(MultiValueRange::try_from(value)?))
+                }
+            }
+        )+
+    }
+}
+
+range_multi_try_from!(UInt8: u8, UInt16: u16, UInt32: u32, UInt64: u64);
```

Comment:
These are all the same, you could consolidate by putting multiple `impl`s in the same macro expansion.

Reply:
Will do. I've actually already picked that approach up from reading your examples and once I spoke macro rules fluently enough it became natural to mimic.